signature-aggregator/aggregator/aggregator_test.go

```diff
@@ -479,3 +479,196 @@ func TestUnmarshalResponse(t *testing.T) {
+						ValidationID: &validationID2,
+						Balance:      &underFunded,
+					},
+				},
+			},
+			connected: &peers.ConnectedCanonicalValidators{
+				ValidatorSet: warp.CanonicalValidatorSet{
+					Validators: []*warp.Validator{
+						{NodeIDs: []ids.NodeID{nodeID1}},
+						{NodeIDs: []ids.NodeID{nodeID2}},
+					},
+				},
+			},
+			excludedIdx: map[int]bool{0: true, 1: true},
+		},
+		{
+			name: "all funded",
+			l1Validators: []platformvm.ClientPermissionlessValidator{
+				{
+					ClientStaker: platformvm.ClientStaker{NodeID: nodeID1},
+					ClientL1Validator: platformvm.ClientL1Validator{
+						ValidationID: &validationID1,
+						Balance:      &funded,
+					},
+				},
+				{
+					ClientStaker: platformvm.ClientStaker{NodeID: nodeID2},
+					ClientL1Validator: platformvm.ClientL1Validator{
+						ValidationID: &validationID2,
+						Balance:      &funded,
+					},
+				},
+			},
+			connected: &peers.ConnectedCanonicalValidators{
+				ValidatorSet: warp.CanonicalValidatorSet{
+					Validators: []*warp.Validator{
+						{NodeIDs: []ids.NodeID{nodeID1}},
+						{NodeIDs: []ids.NodeID{nodeID2}},
+					},
+				},
+			},
+			excludedIdx: map[int]bool{0: false, 1: false},
+		},
+		{
+			name: "mixed L1/non-L1",
+			l1Validators: []platformvm.ClientPermissionlessValidator{
+				{
+					ClientStaker: platformvm.ClientStaker{NodeID: nodeID1},
+					ClientL1Validator: platformvm.ClientL1Validator{
+						ValidationID: &validationID1,
+						Balance:      &funded,
+					},
+				},
+				{
+					// non-L1
+					ClientStaker: platformvm.ClientStaker{
+						NodeID: nodeID2,
+					},
+				},
+			},
+			connected: &peers.ConnectedCanonicalValidators{
+				ValidatorSet: warp.CanonicalValidatorSet{
+					Validators: []*warp.Validator{
+						{NodeIDs: []ids.NodeID{nodeID1}},
+						{NodeIDs: []ids.NodeID{nodeID2}},
+					},
+				},
+			},
+			excludedIdx: map[int]bool{0: false, 1: false},
+		},
+		{
+			name: "nil balance",
+			l1Validators: []platformvm.ClientPermissionlessValidator{
+				{
+					ClientStaker: platformvm.ClientStaker{NodeID: nodeID1},
+					ClientL1Validator: platformvm.ClientL1Validator{
+						ValidationID: &validationID1,
+						Balance:      nil,
+					},
+				},
+			},
+			connected: &peers.ConnectedCanonicalValidators{
+				ValidatorSet: warp.CanonicalValidatorSet{
+					Validators: []*warp.Validator{
+						{NodeIDs: []ids.NodeID{nodeID1}},
+					},
+				},
+			},
+			excludedIdx: map[int]bool{0: true},
+		},
+		{
+			name: "multiple nodeIDs per validator",
+			l1Validators: []platformvm.ClientPermissionlessValidator{
+				{
+					ClientStaker: platformvm.ClientStaker{NodeID: nodeID1},
+					ClientL1Validator: platformvm.ClientL1Validator{
+						ValidationID: &validationID1,
+						Balance:      &funded},
+				},
+				{
+					ClientStaker: platformvm.ClientStaker{NodeID: nodeID2},
+					ClientL1Validator: platformvm.ClientL1Validator{
+						ValidationID: &validationID2,
+						Balance:      &funded},
```

Comment:
May make sense to have one of the validation IDs be `underfunded` here, to confirm that having at least 1 "funded" validation ID is sufficient